### ISSUE TITLE
Ingress interface

### ISSUE THREAD

**Issue by josephglanville** [SeldonIO/seldon-core#1778]
As discussed in #1754 an interface for supporting ingress options in an extensible way would be desired.

In my mind there are 2 main ways ingress controllers in Kubernetes work currently, Custom Resource Definitions (CRDS) and Ingress and/or Service annotations.

Right now Seldon Core supports Istio and Ambassador which cover most of this space already.
Istio support relies on the creation of Istio specific CRDs (VirtualService and DestinationRule) and Ambassador support annotates the Service objects already created as part of a Seldon ML deployment.

Some options like Contour require the capability to do both of these things (as stuff like the upstream protocol for a Contour HTTPProxy CRD is actually derived from annotations on the upstream Service).

As such I think an interface similar to the following would serve pretty well:
```
type Ingress interface {
    AddToScheme(scheme *runtime.Scheme)
    // Takes the manager, does any initial setup and returns a list of objects that should be owned by the constructed controller
    SetupWithManager(mgr ctrl.Manager) ([]runtime.Object, error)
    // Generate predictor resources
    GeneratePredictorResources(mlDep *v1.SeldonDeployment,
        seldonId string,
        namespace string,
        ports []httpGrpcPorts,
        httpAllowed bool,
        grpcAllowed bool) (map[string][]runtime.Object, error)
    // Generate explainer resources
    GenerateExplainerResources(pSvcName string, p *v1.PredictorSpec,
        mlDep *v1.SeldonDeployment,
        seldonId string,
        namespace string,
        engineHttpPort int,
        engineGrpcPort int) (map[string][]runtime.Object, error)
    // Create ingress resources previously generated, returns boolean to indicate readiness
    CreateResources(resources map[string][]runtime.Object, instance *v1.SeldonDeployment, log logr.Logger) (bool, error)
    // Supplies annotations to service objects if required for ingress plugin
    GenerateServiceAnnotations() map[string]string
}
```

This was constructed mainly just by extracting existing functions from Istio/Ambassador support with minimal cleanup. I think we can do a lot better.
Namely we still want to preserve the 2 phase operation of the reconciler. That is the construction and assembly of all the resources (components in the code) before actually invoking the k8s API.

Doing so however requires either passing back resources generically into the reconciler or making the ingress plugins more stateful in some way, or making them contextual so they carry state of a single iteration of the reconciler.

/cc @cliveseldon

**Comment by cliveseldon** [comment]
Thanks @josephglanville  I missed this and it looks very interesting. 

Can you create a design doc to expand on these ideas and options?
We probably need to list all the popular ingress' and how people manage them.
I'm wondering why there isn't a generic k8s project to do this rather than having to write this in an ML focused project. My preference would ideally to be able to extract this to another projects' responsibility.

**Comment by josephglanville** [comment]
Agreed, it would be best if there was a way to abstract these things but unfortunately there isn't a generic interface. It seems the interfaces is a large part of what differentiates the various ingress options as many of them rely on Envoy for their dataplane anyway. This is unfortunate but hopefully there will be consolidation in the future to reduce fragmentation.

I haven't went as far as a design doc but I have pushed up a WIP PR that gets started with something concrete for discussion.

I can't commit to a full analysis of all competing ingress options as a) they are numerous and b) I am short on the time to properly dedicate to such an effort.

That said I think it's fair to say all hold to the description above, i.e they are either annotation based or CRD based so as long as the interface can cater to both of these (we already have one of each and I am looking to introduce a second CRD based option) then it should be sufficiently flexible.

**Comment by cliveseldon** [comment]
If you can try this in [v2](https://docs.seldon.io/projects/seldon-core/en/v2/) which is agnostic to service mesh with a single SVC entrypoint.
